src/libDirectoryService/FinalBlockPreProcessing.cpp

```diff
@@ -97,6 +97,29 @@ void DirectoryService::ExtractDataFromMicroblocks(
                 << microblockTrieRoot.hex());
 }
 
+MBInfoHash DirectoryService::CalculateMBInfoHash(
+    const std::vector<uint32_t>& shardIds,
+    const std::vector<bool>& isMicroBlockEmpty) {
+  vector<unsigned char> mbInfoVec;
+  unsigned int offset = 0;
+  for (const auto& id : shardIds) {
+    Serializable::SetNumber<uint32_t>(mbInfoVec, offset, id, sizeof(uint32_t));
+    offset += sizeof(uint32_t);
+  }
+  for (const auto& isEmpty : isMicroBlockEmpty) {
+    mbInfoVec.push_back(isEmpty ? BoolSign::ISTRUE : BoolSign::ISFALSE);
+    offset += sizeof(unsigned char);
```

Comment:
Other hashes are generated using libMessage (e.g., `GetDSCommitteeHash()`), so it might be better to move this function into libMessage as well.

Reply:
ok will do

Reply:
implemented